plugins/ReactTestComponent.js

```diff
@@ -0,0 +1,61 @@
+var reactTestInstance = Symbol.for('react.test.json');
+
+function handleChildren(result, children, indentation, print, indent) {
+  if (children) {
+    children.forEach(function(child) {
+      result.push('\n');
+      result.push(objectToJSX(child, indentation, print, indent));
+    });
+    result.push('\n');
+  }
+}
+
+function handleProps(result, node, indentation, print, indent) {
+  var props = node.props;
+  if (props) {
+    var indentOpts = {
+      indent: 2 * (indentation + 1)
+    };
+    Object.keys(props).forEach(function(prop) {
+      result.push('\n', indent(prop, indentOpts), '=');
+      var value = props[prop];
+      if (typeof value === 'string') {
+        result.push('"', value, '"');
+      } else {
+        var formatted = indent(print(value), {indent: 2 * (indentation + 2)});
+        result.push('{\n', formatted, '\n', indent('}', indentOpts));
+      }
+    });
+  }
+}
+
+function objectToJSX(root, indentation, print, indent) {
+  indentation = indentation || 0;
+  var indentationOpts = {indent: indentation * 2};
```

Comment:
I find it easier to just keep re-indenting things as they get deeper rather than trying to maintain the state of it. I think you're making this needlessly complicated.